taxdata/puf/impute_qbid_w2.py

```diff
@@ -0,0 +1,87 @@
+"""
+Impute W-2 wages share in Qualified Business Income in PUF.
+
+The objective of this imputation is to improve the accuracy of Qualified
+Business Income Deduction (QBID) revenue estimates produced by Tax-Calculator.
+In the absence of a valid W-2 wage share variable (e.g., when it is set to zero),
+Tax-Calculatorgenerates QBID values that diverge significantly from benchmark
+estimates published by agencies such as the Congressional Budget Office (CBO).
+
+Since no microdata source contains reliable information on the W-2 wage share
+associated with pass-through income, a statistical imputation is required.
+We calibrate the imputation to match CBO's published QBID revenue estimates.
+To be noticed, there is an observed inconsistency on the CBO published QBID
+estimates between the years before 2020 and the years after 2021. So only the
+QBID value from the years 2021 ~ 2025 are selected targets, which would better
+reflect the policy environments of the recent years. Targets will be updated as
+new CBO QBID projections become available.
+
+The imputation strategy is to minimize the distance between Tax-Calculator's
+simulated QBID revenue and CBO's published targets through an optimization
+process. Specifically, we define a loss function to measure the year-by-year
+squared deviations in aggregate QBID amounts and solve for the W-2 wage share
+parameter which minimizes this loss.
+
+The imputation will add the PT_w2_binc_wages into puf.csv file.
+"""
+
+import pandas as pd
+import taxcalc as tc
+import numpy as np
+from scipy.optimize import minimize_scalar
+from functools import partial
+
+
+def impute_PT_binc_w2_wages(data):
+    """
+    Main function in impute_qbid_w2.py file.
+    Argument: puf.csv DataFrame just before imputation is done.
+    Returns: puf.csv DataFrame with imputed W-2 wage share variable.
+    """
+    # add the qualified business income variable, to be noticed the self employment 
+    # part is not included, to simplyfy calculation.
+    data["qbinc"] = data["e00900"] + data["e26270"] + data["e02100"] + data["e27200"]
+    # solve for W-2 wage share in QBI
+    w2_ratio = opt_ratio(data)
+    data["PT_binc_w2_wages"] = w2_ratio * data["qbinc"]
+    del data["qbinc"]
+    return data
+
+
+def qbid_value(data, ratio):
+    """
+    Function that calculate the QBID aggregates through years.
+    """
+    qbided = []
+    df = data.copy()
+    df["PT_binc_w2_wages"] = ratio * df["qbinc"]
+    pol = tc.Policy()
+    recs = tc.Records(data=df)
+    calc0 = tc.Calculator(policy=pol, records=recs)
+    for year in range(2021, 2026):
```

Comment:
@bodiyang Should the year range be an argument here? Any way to make it a function of QBID parameters in Tax-Calculator's current law baseline?  Or why not just go through the budget window used in Tax-Calculator?

Reply:
> @bodiyang Should the year range be an argument here? Any way to make it a function of QBID parameters in Tax-Calculator's current law baseline? Or why not just go through the budget window used in Tax-Calculator?

Because I only wanted to select the year 2021 ~ 2025 QBID value as targets, not other years. It is noticed that there is a gap between the years before 2020 and the years after 2021 in CBO's published QBID aggregates. So the year 2021 and after probably better reflects the recent years QBID trajectory. Years before 2020 is a distraction for the model's optimization process.

Reply:
Ok - understood. Thanks!